### PR TITLE
Guided Tours: `trackClicks` higher-order component

### PR DESCRIPTION
**Pull request by mcsf**
### Experimental

A higher-order component — like `translate` or `connect` — that allows clicks to a given wrapped component to be tracked via the Redux dispatcher.

This aims to solve current limitations with the way Guided Tours listens to DOM elements for step transitions, and, if deemed adequate, it could eventually simplify the view-layer logic.

### In a nutshell

The following:

```diff
-export default ThemeMoreButton;
+export default tracked(
+	( props ) => ( { themeId: props.theme.id } )
+)( ThemeMoreButton );
```

yields:

gt-tracked

### Questions

- [x] Move to its own module, decouple from Guided Tours?
- [ ] In that case, what should the `type` of the dispatched action be?
- [ ] Reuse `data-tip-target` attribute or analogous?

```js
// selectively dispatch click action by
// bailing if no action object returned
trackClicks(
  props => props[ 'data-tip-target' ]
    ? { type, target: props[ 'data-tip-target' ] }
    : undefined
)( Component )
```

### Next

- [ ] Instead of just `onClick`, allow arbitrary events to be tracked and mapped to actions?
- [x] Allow passing specific action type for the `dispatch` call?

Test live: https://calypso.live/?branch=add/guided-tours-tracked-hoc

### CODE REVIEW

File: client/layout/guided-tours/tracked.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import storeShape from 'react-redux/lib/utils/storeShape';
+import { noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { GUIDED_TOUR_TRACKED_CLICK } from 'state/action-types.js';
+
+export default (
+		mapPropsToAction = defaultMapPropsToAction
```
Comment (mcsf) [inline client/layout/guided-tours/tracked.js]: Should this instead be named something like `selectProps`, `selectedTrackedProps`? We're not actually mapping to an action, as the component itself will later add `type`.

**Comment by mcsf** [comment]
This actually breaks tests for `components/theme` due to mocking of the More button. I'll give this some thought.

**Comment by mcsf** [comment]
Looping @seear and @ockham in, just for awareness. :)

**Comment by seear** [comment]
@mcsf: the changes to `client/components/theme/test/index.jsx` look perfectly tidy to me, unless I'm missing something?

**Comment by mcsf** [comment]
> the changes to client/components/theme/test/index.jsx look perfectly tidy to me, unless I'm missing something?

@seear: cool, thanks. I wanted a sanity check, since the whole grabbing of the `default` value from the module and then object-assigning the mocked component into a copy of the module seemed a bit original.

**Comment by seear** [comment]
> grabbing of the default value from the module and then object-assigning the mocked component into a copy of the module

Seems to fit the principle of mocking as little as necessary 😄

**Comment by mcsf** [comment]
Closing this one in favor of #7304

**Pull request closed without merging**